test/bigquery.ts

```diff
@@ -1332,6 +1411,79 @@ describe('BigQuery', () => {
             assert.ifError
           );
         });
+
+        it('should allow for optional parameter types', done => {
+          bq.createJob = (reqOpts: JobOptions) => {
+            // tslint:disable-next-line no-any
+            assert.strictEqual((reqOpts as any).params, undefined);
+            done();
+          };
+
+          bq.createQueryJob(
+            {
+              query: QUERY_STRING,
+              params: POSITIONAL_PARAMS,
+              types: POSITIONAL_TYPES,
+            },
+            assert.ifError
+          );
+        });
+
+        it('should throw for invalid provided types', done => {
+          bq.createJob = (reqOpts: JobOptions) => {
+            done();
+          };
+
+          const INVALID_TYPES = ['invalid'];
+
+          assert.throws(() => {
+            bq.createQueryJob({
+              query: QUERY_STRING,
+              params: POSITIONAL_PARAMS,
+              types: INVALID_TYPES,
+            });
+          }, /Invalid type provided./);
+        });
+
+        it('should throw for invalid type structure provided', done => {
+          bq.createJob = (reqOpts: JobOptions) => {
+            done();
+          };
+
+          assert.throws(() => {
+            bq.createQueryJob({
+              query: QUERY_STRING,
+              params: POSITIONAL_PARAMS,
+              types: NAMED_TYPES,
+            });
+          }, /Provided types must match the value type passed to `params`/);
+        });
+
+        it('should throw for empty array without provided types', done => {
```

Comment:
This test should only be in the describe block for getTypeDescriptorFromProvidedType_.